fvm/src/kernel/mod.rs

```diff
@@ -21,38 +21,68 @@ use fvm_shared::sys::out::vm::MessageContext;
 use fvm_shared::sys::SendFlags;
 use fvm_shared::{ActorID, MethodNum};
 
+mod blocks;
 mod hash;
 
-mod blocks;
 pub mod default;
+pub mod filecoin;
 
 pub(crate) mod error;
 
+use ambassador::delegatable_trait;
 pub use error::{ClassifyResult, Context, ExecutionError, Result, SyscallError};
 use fvm_shared::event::StampedEvent;
 pub use hash::SupportedHashes;
 use multihash::MultihashGeneric;
+use wasmtime::Linker;
 
 use crate::call_manager::CallManager;
 use crate::gas::{Gas, GasTimer, PriceList};
 use crate::machine::limiter::MemoryLimiter;
 use crate::machine::Machine;
+use crate::syscalls::InvocationData;
 
 pub struct CallResult {
     pub block_id: BlockId,
     pub block_stat: BlockStat,
     pub exit_code: ExitCode,
 }
 
+pub trait ConstructKernel<C> {
+    /// Construct a new [`Kernel`] from the parameterized [`CallManager`].
+    ///
+    /// - `caller` is the ID of the _immediate_ caller.
+    /// - `actor_id` is the ID of _this_ actor.
+    /// - `method` is the method that has been invoked.
+    /// - `value_received` is value received due to the current call.
+    /// - `blocks` is the initial block registry (should already contain the parameters).
+    #[allow(clippy::too_many_arguments)]
+    fn new(
+        mgr: C,
+        blocks: BlockRegistry,
+        caller: ActorID,
+        actor_id: ActorID,
+        method: MethodNum,
+        value_received: TokenAmount,
+        read_only: bool,
+    ) -> Self
+    where
+        Self: Sized;
+}
+
 /// The "kernel" implements the FVM interface as presented to the actors. It:
 ///
 /// - Manages the Actor's state.
 /// - Tracks and charges for IPLD & syscall-specific gas.
 ///
 /// Actors may call into the kernel via the syscalls defined in the [`syscalls`][crate::syscalls]
 /// module.
+#[delegatable_trait]
 pub trait Kernel:
-    ActorOps
+    SyscallHandler<<Self as Kernel>::Kernel>
```

Comment:
Do we need this? IMO:

1. This really should be the other way around (kernels are syscall handlers).
2. I'm not sure if we even need to care about that either. We can likely treat them as different things.


So, I think the `SyscallHandler` implementation changes to:

```rust
impl<K, T> SyscallHandler<K> for DefaultFilecoinKernel<T>
where
    K: Kernel,
    T: SyscallHandler<Self>
```

Reply:
Are you suggesting removing SyscallHandler from the Kernel? How would we call the `bind_syscalls` on it when initializing the engine then, or am I misunderstanding?

Reply:
I was... but yeah, that might be even more annoying (we'd have to change all `: Kernel` constraints to `: SyscallHandler` or `: Kernel + SyscallHandler`. So, maybe later but not now.